File: app/src/main/java/com/example/paetz/yacguide/network/NetworkTask.java
package com.example.paetz.yacguide.network;

import android.os.AsyncTask;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class NetworkTask extends AsyncTask<String, Void, String> {

    private int _requestId;
    private NetworkListener _listener;

    public NetworkTask(int requestId, NetworkListener listener) {
        _requestId = requestId;
        _listener = listener;
    }

    @Override
    protected String doInBackground(String... urls) {
        if (urls.length != 1) {
            System.err.println("Can only process one url");
            return null;
        }
        HttpURLConnection connection = null;
        String result = null;
        try {
            connection = (HttpURLConnection) new URL(urls[0]).openConnection();
            InputStream inStream = new BufferedInputStream(connection.getInputStream());
            ByteArrayOutputStream resStream = new ByteArrayOutputStream();
            byte[] buffer = new byte[1024];
            int length;
            while ((length = inStream.read(buffer)) != -1) {
                resStream.write(buffer, 0, length);
            }
            result = resStream.toString("UTF-8");
        } catch (IOException e) {
            System.err.println(e);
        } finally {
            if (connection != null) {
                connection.disconnect();
            }
        }
        return result;
    }

    @Override
    protected void onPostExecute(String result) {
        _listener.onNetworkTaskResolved(_requestId, result);
    }
}

File: app/src/main/java/com/yacgroup/yacguide/network/NetworkTask.kt
/*
 * Copyright (C) 2019, 2022, 2023 Axel Paetzold
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.yacgroup.yacguide.network

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

class NetworkTask(private val _request: NetworkRequest,
                  private val _scope: CoroutineScope,
                  private val _listener: NetworkListener) {

    fun execute() = _scope.launch {
        val result = _performNetworkRequest()
        _listener.onNetworkTaskResolved(NetworkAnswer(_request, result))
    }

    private suspend fun _performNetworkRequest(): String = withContext(Dispatchers.IO) {
        val connection = URL(_request.url).openConnection() as HttpURLConnection

        return@withContext try {
            connection.inputStream.bufferedReader().readText()
        } catch (e: IOException) {
            System.err.println(e)
            ""
        } finally {
            connection.disconnect()
        }
    }
}
